**Linux** [Build 10634](https://ci.wildfly.org/viewLog.html?buildId=29866&buildTypeId=WFPR) outcome was **FAILURE** using a merge of 21da4459c4dcbe4256e44222c196f5036d025bfc
Summary: Tests failed: 1 (1 new), passed: 3523, ignored: 183 Build time: 01:08:38
### Failed tests
```
org.jboss.as.test.integration.jmx.ModelControllerMBeanTestCase.testAllMBeanInfos: java.lang.AssertionError: {jboss.as:subsystem=security-manager=java.io.IOException: Internal server error.}
	at org.jboss.as.test.integration.jmx.ModelControllerMBeanTestCase.testAllMBeanInfos(ModelControllerMBeanTestCase.java:92)
------- Stdout: -------
11:50:55,371 WARN  [org.jboss.remotingjmx.Util] (main) The protocol 'http-remoting-jmx' is deprecated, instead you should use 'remote+http'.
 [33m11:50:58,634 WARN  [org.jboss.remotingjmx.protocol.v2.ServerCommon] (pool-207-thread-1) Unexpected internal error: java.lang.IllegalArgumentException
	at org.jboss.dmr.ModelValue.getChild(ModelValue.java:115)
	at org.jboss.dmr.PropertyModelValue.getChild(PropertyModelValue.java:120)
	at org.jboss.dmr.ModelNode.get(ModelNode.java:861)
	at org.jboss.as.jmx.model.TypeConverters$ComplexTypeConverter.fromModelNode(TypeConverters.java:539)
	at org.jboss.as.jmx.model.TypeConverters$ListTypeConverter.fromModelNode(TypeConverters.java:445)
	at org.jboss.as.jmx.model.TypeConverters.fromModelNode(TypeConverters.java:115)
	at org.jboss.as.jmx.model.MBeanInfoFactory.getIfExists(MBeanInfoFactory.java:316)
	at org.jboss.as.jmx.model.MBeanInfoFactory.getParameterInfos(MBeanInfoFactory.java:276)
	at org.jboss.as.jmx.model.MBeanInfoFactory.getOperation(MBeanInfoFactory.java:243)
	at org.jboss.as.jmx.model.MBeanInfoFactory.addChildAddOperations(MBeanInfoFactory.java:237)
	at org.jboss.as.jmx.model.MBeanInfoFactory.getOperations(MBeanInfoFactory.java:226)
	at org.jboss.as.jmx.model.MBeanInfoFactory.createMBeanInfo(MBeanInfoFactory.java:141)
	at org.jboss.as.jmx.model.MBeanInfoFactory.createMBeanInfo(MBeanInfoFactory.java:133)
	at org.jboss.as.jmx.model.ModelControllerMBeanHelper.getMBeanInfo(ModelControllerMBeanHelper.java:259)
	at org.jboss.as.jmx.model.ModelControllerMBeanServerPlugin.getMBeanInfo(ModelControllerMBeanServerPlugin.java:170)
	at org.jboss.as.jmx.PluggableMBeanServerImpl.getMBeanInfo(PluggableMBeanServerImpl.java:582)
	at org.jboss.as.jmx.PluggableMBeanServerImpl.getMBeanInfo(PluggableMBeanServerImpl.java:570)
	at org.jboss.as.jmx.BlockingNotificationMBeanServer.getMBeanInfo(BlockingNotificationMBeanServer.java:251)
	at org.jboss.remotingjmx.protocol.v2.ServerProxy$GetMBeanInfoHandler.handle(ServerProxy.java:778)
	at org.jboss.remotingjmx.protocol.v2.ServerCommon$MessageReciever$1$1.run(ServerCommon.java:153)
	at org.jboss.as.jmx.ServerInterceptorFactory$Interceptor$1.run(ServerInterceptorFactory.java:75)
	at org.jboss.as.jmx.ServerInterceptorFactory$Interceptor$1.run(ServerInterceptorFactory.java:70)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.jboss.as.controller.AccessAuditContext.doAs(AccessAuditContext.java:149)
	at org.jboss.as.jmx.ServerInterceptorFactory$Interceptor.handleEvent(ServerInterceptorFactory.java:70)
	at org.jboss.remotingjmx.protocol.v2.ServerCommon$MessageReciever$1.run(ServerCommon.java:149)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

 [0m


```

--- a/security-manager/src/main/java/org/wildfly/extension/security/manager/DeploymentPermissionsResourceDefinition.java
+++ b/security-manager/src/main/java/org/wildfly/extension/security/manager/DeploymentPermissionsResourceDefinition.java
@@ -44,6 +44,7 @@ import org.jboss.as.controller.ReloadRequiredRemoveStepHandler;
 import org.jboss.as.controller.SimpleAttributeDefinition;
 import org.jboss.as.controller.SimpleAttributeDefinitionBuilder;
 import org.jboss.as.controller.registry.AttributeAccess;
+import org.jboss.dmr.ModelNode;
 import org.jboss.dmr.ModelType;
 
 /**
@@ -78,6 +79,17 @@ class DeploymentPermissionsResourceDefinition extends PersistentResourceDefiniti
 
     private static final ObjectTypeAttributeDefinition PERMISSIONS_VALUE_TYPE =
             ObjectTypeAttributeDefinition.Builder.of(PERMISSION, CLASS, NAME, ACTIONS, MODULE).build();
+
+    // this was supposed to be the attribute definition's default value, but using it as a default results in an error when
+    // trying to obtain the MBeanInfo for the security manager subsystem because default lists are not supported. The error
+    // message 'Default value not supported for ArrayType and TabularType' is displayed in an exception thrown by the
+    // javax.management.openmbean.OpenMBeanInfoSupport class.
+    static final ModelNode DEFAULT_MAXIMUM_SET;
+    static {
+        final ModelNode defaultPermission = new ModelNode();
+        defaultPermission.get(PERMISSION_CLASS).set("java.security.AllPermission");
+        DEFAULT_MAXIMUM_SET = new ModelNode().add(defaultPermission);
+    };
 
     static final AttributeDefinition MAXIMUM_PERMISSIONS =
             ObjectListAttributeDefinition.Builder.of(Constants.MAXIMUM_PERMISSIONS, PERMISSIONS_VALUE_TYPE)

--- a/security-manager/src/main/java/org/wildfly/extension/security/manager/Namespace.java
+++ b/security-manager/src/main/java/org/wildfly/extension/security/manager/Namespace.java
@@ -35,12 +35,9 @@ enum Namespace {
 
     UNKNOWN(null),
 
-    SECURITY_MANAGER_1_0("urn:jboss:domain:security-manager:1.0");
+    SECURITY_MANAGER_1_0("urn:jboss:domain:security-manager:1.0"),
 
-    /**
-     * The current namespace version.
-     */
-    public static final Namespace CURRENT = SECURITY_MANAGER_1_0;
+    SECURITY_MANAGER_3_0("urn:jboss:domain:security-manager:3.0");
 
     private final String name;
 

--- a/security-manager/src/main/java/org/wildfly/extension/security/manager/SecurityManagerExtension.java
+++ b/security-manager/src/main/java/org/wildfly/extension/security/manager/SecurityManagerExtension.java
@@ -25,15 +25,24 @@ package org.wildfly.extension.security.manager;
 
 import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.SUBSYSTEM;
 
+import java.util.Map;
+
 import org.jboss.as.controller.Extension;
 import org.jboss.as.controller.ExtensionContext;
 import org.jboss.as.controller.ModelVersion;
+import org.jboss.as.controller.PathAddress;
 import org.jboss.as.controller.PathElement;
 import org.jboss.as.controller.SubsystemRegistration;
 import org.jboss.as.controller.descriptions.StandardResourceDescriptionResolver;
 import org.jboss.as.controller.operations.common.GenericSubsystemDescribeHandler;
 import org.jboss.as.controller.parsing.ExtensionParsingContext;
 import org.jboss.as.controller.registry.ManagementResourceRegistration;
+import org.jboss.as.controller.transform.TransformationContext;
+import org.jboss.as.controller.transform.description.RejectAttributeChecker;
+import org.jboss.as.controller.transform.description.ResourceTransformationDescriptionBuilder;
+import org.jboss.as.controller.transform.description.TransformationDescription;
+import org.jboss.dmr.ModelNode;
+import org.wildfly.extension.security.manager.logging.SecurityManagerLogger;
 
 /*
  * This class implements the security manager extension.
@@ -45,11 +54,12 @@ public class SecurityManagerExtension implements Extension {
     public static final PathElement SUBSYSTEM_PATH = PathElement.pathElement(SUBSYSTEM, Constants.SUBSYSTEM_NAME);
     protected static final String RESOURCE_NAME = SecurityManagerExtension.class.getPackage().getName() + ".LocalDescriptions";
 
-    private static final int MANAGEMENT_API_MAJOR_VERSION = 2;
+    private static final int MANAGEMENT_API_MAJOR_VERSION = 3;
     private static final int MANAGEMENT_API_MINOR_VERSION = 0;
     private static final int MANAGEMENT_API_MICRO_VERSION = 0;
     private static final ModelVersion CURRENT_MODEL_VERSION =
             ModelVersion.create(MANAGEMENT_API_MAJOR_VERSION, MANAGEMENT_API_MINOR_VERSION, MANAGEMENT_API_MICRO_VERSION);
+    private static final ModelVersion EAP_7_0_0_MODEL_VERSION = ModelVersion.create(2, 0, 0);
 
     public static StandardResourceDescriptionResolver getResolver(final String... keyPrefix) {
         StringBuilder prefix = new StringBuilder(Constants.SUBSYSTEM_NAME);
@@ -64,14 +74,42 @@ public class SecurityManagerExtension implements Extension {
         final SubsystemRegistration subsystem = context.registerSubsystem(Constants.SUBSYSTEM_NAME, CURRENT_MODEL_VERSION);
         final ManagementResourceRegistration registration = subsystem.registerSubsystemModel(SecurityManagerRootDefinition.INSTANCE);
         registration.registerOperationHandler(GenericSubsystemDescribeHandler.DEFINITION, GenericSubsystemDescribeHandler.INSTANCE, false);
-        subsystem.registerXMLElementWriter(SecurityManagerSubsystemParser.INSTANCE);
+        subsystem.registerXMLElementWriter(SecurityManagerSubsystemParser_3_0.INSTANCE);
 
-        //Since version 1.0.0 of this subsystem has not been part of any product release I am not registering
-        //any transformers for it
+        // register transformers for released EAP versions that include the security manager subsystem.
+        registerTransformers_EAP_7_0_0(subsystem);
     }
 
     @Override
     public void initializeParsers(final ExtensionParsingContext context) {
-        context.setSubsystemXmlMapping(Constants.SUBSYSTEM_NAME, Namespace.SECURITY_MANAGER_1_0.getUriString(), SecurityManagerSubsystemParser.INSTANCE);
+        context.setSubsystemXmlMapping(Constants.SUBSYSTEM_NAME, Namespace.SECURITY_MANAGER_1_0.getUriString(), SecurityManagerSubsystemParser_1_0.INSTANCE);
+        context.setSubsystemXmlMapping(Constants.SUBSYSTEM_NAME, Namespace.SECURITY_MANAGER_3_0.getUriString(), SecurityManagerSubsystemParser_3_0.INSTANCE);
+    }
+
+    /**
+     * Registers the transformers for JBoss EAP 7.0.0.
+     *
+     * @param subsystemRegistration a reference to the {@link SubsystemRegistration}.
+     */
+    private void registerTransformers_EAP_7_0_0(final SubsystemRegistration subsystemRegistration) {
+        ResourceTransformationDescriptionBuilder builder = ResourceTransformationDescriptionBuilder.Factory.createSubsystemInstance();
+        builder.addChildResource(DeploymentPermissionsResourceDefinition.DEPLOYMENT_PERMISSIONS_PATH).
+                getAttributeBuilder().addRejectCheck(new RejectAttributeChecker.DefaultRejectAttributeChecker() {
+
+            @Override
+            protected boolean rejectAttribute(PathAddress address, String attributeName, ModelNode value, TransformationContext context) {
+                // reject the maximum set if it is defined and empty as that would result in complete incompatible policies
+                // being used in nodes running earlier versions of the subsystem.
+                if (value.isDefined() && value.asList().isEmpty())
+                    return true;
+                return false;
+            }
+
+            @Override
+            public String getRejectionLogMessage(Map<String, ModelNode> attributes) {
+                return SecurityManagerLogger.ROOT_LOGGER.rejectedEmptyMaximumSet();
+            }
+        }, DeploymentPermissionsResourceDefinition.MAXIMUM_PERMISSIONS);
+        TransformationDescription.Tools.register(builder.build(), subsystemRegistration, EAP_7_0_0_MODEL_VERSION);
     }
 }

--- a/security-manager/src/main/java/org/wildfly/extension/security/manager/SecurityManagerSubsystemAdd.java
+++ b/security-manager/src/main/java/org/wildfly/extension/security/manager/SecurityManagerSubsystemAdd.java
@@ -23,15 +23,18 @@
 
 package org.wildfly.extension.security.manager;
 
-import static org.wildfly.extension.security.manager.Constants.DEFAULT_VALUE;
-import static org.wildfly.extension.security.manager.Constants.DEPLOYMENT_PERMISSIONS;
-import static org.wildfly.extension.security.manager.Constants.MAXIMUM_PERMISSIONS;
-import static org.wildfly.extension.security.manager.Constants.MINIMUM_PERMISSIONS;
 import static org.wildfly.extension.security.manager.Constants.PERMISSION_ACTIONS;
 import static org.wildfly.extension.security.manager.Constants.PERMISSION_MODULE;
 import static org.wildfly.extension.security.manager.Constants.PERMISSION_NAME;
+import static org.wildfly.extension.security.manager.DeploymentPermissionsResourceDefinition.ACTIONS;
+import static org.wildfly.extension.security.manager.DeploymentPermissionsResourceDefinition.CLASS;
+import static org.wildfly.extension.security.manager.DeploymentPermissionsResourceDefinition.DEFAULT_MAXIMUM_SET;
+import static org.wildfly.extension.security.manager.DeploymentPermissionsResourceDefinition.DEPLOYMENT_PERMISSIONS_PATH;
+import static org.wildfly.extension.security.manager.DeploymentPermissionsResourceDefinition.MAXIMUM_PERMISSIONS;
+import static org.wildfly.extension.security.manager.DeploymentPermissionsResourceDefinition.MINIMUM_PERMISSIONS;
+import static org.wildfly.extension.security.manager.DeploymentPermissionsResourceDefinition.MODULE;
+import static org.wildfly.extension.security.manager.DeploymentPermissionsResourceDefinition.NAME;
 
-import java.security.AllPermission;
 import java.security.Permission;
 import java.util.ArrayList;
 import java.util.List;
@@ -50,7 +53,6 @@ import org.jboss.modules.Module;
 import org.jboss.modules.ModuleIdentifier;
 import org.jboss.modules.ModuleLoadException;
 import org.jboss.modules.security.FactoryPermissionCollection;
-import org.jboss.modules.security.ImmediatePermissionFactory;
 import org.jboss.modules.security.LoadedPermissionFactory;
 import org.jboss.modules.security.PermissionFactory;
 import org.wildfly.extension.security.manager.deployment.PermissionsParserProcessor;
@@ -60,7 +62,7 @@ import org.wildfly.security.manager.WildFlySecurityManager;
 
 /**
  * Handler that adds the security manager subsystem. It instantiates the permissions specified in the subsystem configuration
- * and installs the service that will activate the {@link WildFlySecurityManager}.
+ * and installs the DUPs that parse and validate the deployment permissions.
  *
  * @author <a href="sguilhen@jboss.com">Stefan Guilhen</a>
  */
@@ -85,7 +87,7 @@ class SecurityManagerSubsystemAdd extends AbstractBoottimeAddStepHandler {
             public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
                 final Resource resource = context.readResource(PathAddress.EMPTY_ADDRESS);
                 ModelNode node = Resource.Tools.readModel(resource);
-                launchServices(context, node);
+                installProcessors(context, node);
                 // Rollback handled by the parent step
                 context.completeStep(OperationContext.RollbackHandler.NOOP_ROLLBACK_HANDLER);
             }
@@ -93,42 +95,39 @@ class SecurityManagerSubsystemAdd extends AbstractBoottimeAddStepHandler {
     }
 
     /**
-     * Retrieves the permissions configured in the security manager subsystem and installs the service that will enable
-     * the {@link WildFlySecurityManager}.
+     * Retrieves the permissions configured in the security manager subsystem and installs the DUPs that parse and validate
+     * the deployment permissions.
      *
      * @param context a reference to the {@link OperationContext}.
-     * @param node the {@link ModelNode} that contains all the configured permissions.
-     *                   be added.
-     * @throws OperationFailedException if an error occurs while launching the security manager services.
+     * @param node the {@link ModelNode} that contains all the configured permissions be added.
+     * @throws OperationFailedException if an error occurs while processing the permissions specified in the subsystem.
      */
-    protected void launchServices(final OperationContext context, final ModelNode node)
+    protected void installProcessors(final OperationContext context, final ModelNode node)
             throws OperationFailedException {
 
         // get the minimum set of deployment permissions.
-        final List<PermissionFactory> minimumSet = this.retrievePermissionSet(context,
-                this.peek(node, DEPLOYMENT_PERMISSIONS, DEFAULT_VALUE, MINIMUM_PERMISSIONS));
+        final ModelNode deploymentPermissionsModel = node.get(DEPLOYMENT_PERMISSIONS_PATH.getKeyValuePair());
+        final ModelNode minimumPermissionsNode = MINIMUM_PERMISSIONS.resolveModelAttribute(context, deploymentPermissionsModel);
+        final List<PermissionFactory> minimumSet = this.retrievePermissionSet(context, minimumPermissionsNode);
 
         // get the maximum set of deployment permissions.
-        final List<PermissionFactory> maximumSet = this.retrievePermissionSet(context,
-                this.peek(node, DEPLOYMENT_PERMISSIONS, DEFAULT_VALUE, MAXIMUM_PERMISSIONS));
+        ModelNode maximumPermissionsNode = MAXIMUM_PERMISSIONS.resolveModelAttribute(context, deploymentPermissionsModel);
+        if (!maximumPermissionsNode.isDefined())
+            maximumPermissionsNode = DEFAULT_MAXIMUM_SET;
+        final List<PermissionFactory> maximumSet = this.retrievePermissionSet(context, maximumPermissionsNode);
 
-        if (maximumSet.isEmpty()) {
-            maximumSet.add(new ImmediatePermissionFactory(new AllPermission()));
-        }
-
-        // validate the configured permissions - the mininum set must be implid by the maximum set.
+        // validate the configured permissions - the minimum set must be implied by the maximum set.
         final FactoryPermissionCollection maxPermissionCollection = new FactoryPermissionCollection(maximumSet.toArray(new PermissionFactory[maximumSet.size()]));
         final StringBuilder failedPermissions = new StringBuilder();
         for (PermissionFactory factory : minimumSet) {
             Permission permission = factory.construct();
             if (!maxPermissionCollection.implies(permission)) {
-                failedPermissions.append("\n\t\t" + permission);
+                failedPermissions.append("\n\t\t").append(permission);
             }
         }
         if (failedPermissions.length() > 0) {
             throw SecurityManagerLogger.ROOT_LOGGER.invalidSubsystemConfiguration(failedPermissions);
         }
-
 
         // install the DUPs responsible for parsing and validating security permissions found in META-INF/permissions.xml.
         context.addStep(new AbstractDeploymentChainStep() {
@@ -151,20 +150,20 @@ class SecurityManagerSubsystemAdd extends AbstractBoottimeAddStepHandler {
      */
     protected List<PermissionFactory> retrievePermissionSet(final OperationContext context, final ModelNode node) throws OperationFailedException {
 
-        final List<PermissionFactory> permissions = new ArrayList<PermissionFactory>();
+        final List<PermissionFactory> permissions = new ArrayList<>();
 
         if (node != null && node.isDefined()) {
             for (ModelNode permissionNode : node.asList()) {
-                String permissionClass = DeploymentPermissionsResourceDefinition.CLASS.resolveModelAttribute(context, permissionNode).asString();
+                String permissionClass = CLASS.resolveModelAttribute(context, permissionNode).asString();
                 String permissionName = null;
                 if (permissionNode.hasDefined(PERMISSION_NAME))
-                    permissionName = DeploymentPermissionsResourceDefinition.NAME.resolveModelAttribute(context, permissionNode).asString();
+                    permissionName = NAME.resolveModelAttribute(context, permissionNode).asString();
                 String permissionActions = null;
                 if (permissionNode.hasDefined(PERMISSION_ACTIONS))
-                    permissionActions = DeploymentPermissionsResourceDefinition.ACTIONS.resolveModelAttribute(context, permissionNode).asString();
+                    permissionActions = ACTIONS.resolveModelAttribute(context, permissionNode).asString();
                 String moduleName = null;
                 if(permissionNode.hasDefined(PERMISSION_MODULE)) {
-                    moduleName =  DeploymentPermissionsResourceDefinition.MODULE.resolveModelAttribute(context, permissionNode).asString();
+                    moduleName =  MODULE.resolveModelAttribute(context, permissionNode).asString();
                 }
                 ClassLoader cl = WildFlySecurityManager.getClassLoaderPrivileged(this.getClass());
                 if(moduleName != null) {
@@ -180,25 +179,5 @@ class SecurityManagerSubsystemAdd extends AbstractBoottimeAddStepHandler {
             }
         }
         return permissions;
-    }
-
-
-    /**
-     * Utility method that traverses a {@link ModelNode} according to the provided path. If a valid node is reached, it is
-     * returned.
-     *
-     * @param node the node to be traversed.
-     * @param args a {@code String[]} that forms the path to be traversed.
-     * @return a reference to the {@link ModelNode} that was reached at the end of the path, or {@code null} if path doesn't
-     * lead to a defined node.
-     */
-    protected ModelNode peek(ModelNode node, String... args) {
-        for (String arg : args) {
-            if (!node.hasDefined(arg)) {
-                return null;
-            }
-            node = node.get(arg);
-        }
-        return node;
     }
 }

--- a/security-manager/src/main/java/org/wildfly/extension/security/manager/SecurityManagerSubsystemParser_1_0.java
+++ b/security-manager/src/main/java/org/wildfly/extension/security/manager/SecurityManagerSubsystemParser_1_0.java
@@ -1,0 +1,63 @@
+/*
+ *
+ *  JBoss, Home of Professional Open Source.
+ *  Copyright 2013, Red Hat, Inc., and individual contributors
+ *  as indicated by the @author tags. See the copyright.txt file in the
+ *  distribution for a full listing of individual contributors.
+ *
+ *  This is free software; you can redistribute it and/or modify it
+ *  under the terms of the GNU Lesser General Public License as
+ *  published by the Free Software Foundation; either version 2.1 of
+ *  the License, or (at your option) any later version.
+ *
+ *  This software is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ *  Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public
+ *  License along with this software; if not, write to the Free
+ *  Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ *  02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.wildfly.extension.security.manager;
+
+import org.jboss.as.controller.PersistentResourceXMLDescription;
+import org.jboss.as.controller.PersistentResourceXMLParser;
+
+import static org.jboss.as.controller.PersistentResourceXMLDescription.builder;
+
+/**
+ * This class implements a parser for version 1.0 of the security manager subsystem.
+ *
+ * @author <a href="mailto:sguilhen@redhat.com">Stefan Guilhen</a>
+ */
+class SecurityManagerSubsystemParser_1_0 extends PersistentResourceXMLParser {
+
+    static final SecurityManagerSubsystemParser_1_0 INSTANCE = new SecurityManagerSubsystemParser_1_0();
+
+    private static final PersistentResourceXMLDescription xmlDescription;
+
+    static {
+        xmlDescription = builder(SecurityManagerRootDefinition.INSTANCE, Namespace.SECURITY_MANAGER_1_0.getUriString())
+                .addChild(builder(DeploymentPermissionsResourceDefinition.INSTANCE)
+                                .setXmlElementName(Constants.DEPLOYMENT_PERMISSIONS)
+                                .addAttribute(DeploymentPermissionsResourceDefinition.MINIMUM_PERMISSIONS)
+                                .addAttribute(DeploymentPermissionsResourceDefinition.MAXIMUM_PERMISSIONS)
+                )
+                .build();
+    }
+
+
+    /**
+     * Private constructor to enforce usage of the static singleton instance.
+     */
+    private SecurityManagerSubsystemParser_1_0() {
+    }
+
+    @Override
+    public PersistentResourceXMLDescription getParserDescription() {
+        return xmlDescription;
+    }
+}

--- a/security-manager/src/main/java/org/wildfly/extension/security/manager/SecurityManagerSubsystemParser_3_0.java
+++ b/security-manager/src/main/java/org/wildfly/extension/security/manager/SecurityManagerSubsystemParser_3_0.java
@@ -23,24 +23,24 @@
 
 package org.wildfly.extension.security.manager;
 
+import static org.jboss.as.controller.PersistentResourceXMLDescription.builder;
+
 import org.jboss.as.controller.PersistentResourceXMLDescription;
 import org.jboss.as.controller.PersistentResourceXMLParser;
 
-import static org.jboss.as.controller.PersistentResourceXMLDescription.builder;
-
 /**
- * This class implements a parser for the security manager subsystem.
+ * This class implements a parser for version 3.0 of the security manager subsystem.
  *
  * @author <a href="mailto:sguilhen@redhat.com">Stefan Guilhen</a>
  */
-class SecurityManagerSubsystemParser extends PersistentResourceXMLParser {
+class SecurityManagerSubsystemParser_3_0 extends PersistentResourceXMLParser {
 
-    static final SecurityManagerSubsystemParser INSTANCE = new SecurityManagerSubsystemParser();
+    static final SecurityManagerSubsystemParser_3_0 INSTANCE = new SecurityManagerSubsystemParser_3_0();
 
     private static final PersistentResourceXMLDescription xmlDescription;
 
     static {
-        xmlDescription = builder(SecurityManagerRootDefinition.INSTANCE, Namespace.SECURITY_MANAGER_1_0.getUriString())
+        xmlDescription = builder(SecurityManagerRootDefinition.INSTANCE, Namespace.SECURITY_MANAGER_3_0.getUriString())
                 .addChild(builder(DeploymentPermissionsResourceDefinition.INSTANCE)
                                 .setXmlElementName(Constants.DEPLOYMENT_PERMISSIONS)
                                 .addAttribute(DeploymentPermissionsResourceDefinition.MINIMUM_PERMISSIONS)
@@ -53,7 +53,7 @@ class SecurityManagerSubsystemParser extends PersistentResourceXMLParser {
     /**
      * Private constructor to enforce usage of the static singleton instance.
      */
-    private SecurityManagerSubsystemParser() {
+    private SecurityManagerSubsystemParser_3_0() {
     }
 
     @Override

--- a/security-manager/src/main/java/org/wildfly/extension/security/manager/logging/SecurityManagerLogger.java
+++ b/security-manager/src/main/java/org/wildfly/extension/security/manager/logging/SecurityManagerLogger.java
@@ -77,4 +77,13 @@ public interface SecurityManagerLogger extends BasicLogger {
      */
     @Message(id = 4, value = "Deployment configuration error: the following permissions are not implied by the maximum permissions set %s")
     DeploymentUnitProcessingException invalidDeploymentConfiguration(StringBuilder failedPermissions);
+
+    /**
+     * Creates a message indicating that empty maximum sets are not understood in the target model version and must thus
+     * be rejected.
+     *
+     * @return the constructed {@link String} message.
+     */
+    @Message(id = 5, value = "Empty maximum sets are not understood in the target model version and must be rejected")
+    String rejectedEmptyMaximumSet();
 }

--- a/security-manager/src/main/resources/schema/wildfly-security-manager_3_0.xsd
+++ b/security-manager/src/main/resources/schema/wildfly-security-manager_3_0.xsd
@@ -1,0 +1,101 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!--
+  ~ JBoss, Home of Professional Open Source.
+  ~ Copyright 2011, Red Hat, Inc., and individual contributors
+  ~ as indicated by the @author tags. See the copyright.txt file in the
+  ~ distribution for a full listing of individual contributors.
+  ~
+  ~ This is free software; you can redistribute it and/or modify it
+  ~ under the terms of the GNU Lesser General Public License as
+  ~ published by the Free Software Foundation; either version 2.1 of
+  ~ the License, or (at your option) any later version.
+  ~
+  ~ This software is distributed in the hope that it will be useful,
+  ~ but WITHOUT ANY WARRANTY; without even the implied warranty of
+  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  ~ Lesser General Public License for more details.
+  ~
+  ~ You should have received a copy of the GNU Lesser General Public
+  ~ License along with this software; if not, write to the Free
+  ~ Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema"
+           targetNamespace="urn:jboss:domain:security-manager:1.0"
+           xmlns="urn:jboss:domain:security-manager:1.0"
+           elementFormDefault="qualified"
+           attributeFormDefault="unqualified"
+           version="1.0">
+
+    <!-- The security manager subsystem root element -->
+    <xs:element name="subsystem" type="securityManagerSubsystemType"/>
+
+    <xs:complexType name="securityManagerSubsystemType">
+        <xs:annotation>
+            <xs:documentation>
+                <![CDATA[
+                The security manager subsystem configuration.
+                        ]]>
+            </xs:documentation>
+        </xs:annotation>
+        <xs:sequence>
+            <xs:element name="deployment-permissions" minOccurs="0" maxOccurs="1" type="deploymentPermissionsConfigType"/>
+        </xs:sequence>
+    </xs:complexType>
+
+    <xs:complexType name="deploymentPermissionsConfigType">
+        <xs:annotation>
+            <xs:documentation>
+                <![CDATA[
+                The deploymentPermissionConfigType specifies the elements and attributes that can be used to configure
+                the minimum and maximum set of permissions for deployments.
+
+                - minimum-set: this element allows for the configuration of the minimum set of permissions that are to
+                be granted to all deployments.
+                - maximum-set: this element allows for the configuration of the maximum set of permissions that a deployment
+                can have. This set defines the maximum scope of permissions that can be granted to a deployment and acts
+                as a validation mechanims that prevents deployments from acquiring undesirable permissions.
+                        ]]>
+            </xs:documentation>
+        </xs:annotation>
+        <xs:sequence>
+            <xs:element name="minimum-set" minOccurs="0" maxOccurs="1" type="permissionSetType"/>
+            <xs:element name="maximum-set" minOccurs="0" maxOccurs="1" type="permissionSetType"/>
+        </xs:sequence>
+    </xs:complexType>
+
+    <xs:complexType name="permissionSetType">
+        <xs:annotation>
+            <xs:documentation>
+                <![CDATA[
+                The permissionSetConfigType defines a sequence of permission elements that can be used to specify the
+                permissions that are to be granted to deployments or external jars.
+                        ]]>
+            </xs:documentation>
+        </xs:annotation>
+        <xs:sequence>
+            <xs:element name="permission" minOccurs="0" maxOccurs="unbounded" type="permissionType"/>
+        </xs:sequence>
+    </xs:complexType>
+
+    <xs:complexType name="permissionType">
+        <xs:annotation>
+            <xs:documentation>
+                <![CDATA[
+                The permission specifies the attributes used to declare a security permission.
+
+                * class: the permission fully-qualified class name (e.g. java.util.PropertyPermission).
+                * name: the name  of permission target (e.g. java.* for a PropertyPermission).
+                * actions: the permission actions (e.g. read, write)
+                        ]]>
+            </xs:documentation>
+        </xs:annotation>
+        <xs:attribute name="class" type="xs:string" use="required"/>
+        <xs:attribute name="name" type="xs:string" use="optional"/>
+        <xs:attribute name="actions" type="xs:string" use="optional"/>
+        <xs:attribute name="module" type="xs:string" use="optional"/>
+    </xs:complexType>
+
+</xs:schema>

--- a/security-manager/src/main/resources/schema/wildfly-security-manager_3_0.xsd
+++ b/security-manager/src/main/resources/schema/wildfly-security-manager_3_0.xsd
@@ -23,8 +23,8 @@
   -->
 
 <xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema"
-           targetNamespace="urn:jboss:domain:security-manager:1.0"
-           xmlns="urn:jboss:domain:security-manager:1.0"
+           targetNamespace="urn:jboss:domain:security-manager:3.0"
+           xmlns="urn:jboss:domain:security-manager:3.0"
            elementFormDefault="qualified"
            attributeFormDefault="unqualified"
            version="1.0">
@@ -53,10 +53,21 @@
                 the minimum and maximum set of permissions for deployments.
 
                 - minimum-set: this element allows for the configuration of the minimum set of permissions that are to
-                be granted to all deployments.
+                be granted to all deployments. If this element is not specified or if the element is empty in the
+                subsystem configuration, an empty minimum permission set will be created. The effect in this case is that
+                deployment will not get any permissions other than the ones configured in the deployment itself.
                 - maximum-set: this element allows for the configuration of the maximum set of permissions that a deployment
                 can have. This set defines the maximum scope of permissions that can be granted to a deployment and acts
-                as a validation mechanims that prevents deployments from acquiring undesirable permissions.
+                as a validation mechanism that prevents deployments from acquiring undesirable permissions. If this element
+                is not specified, a default maximum set with a single java.security.AllPermission is created (and it thus
+                implies all permissions specified in the minimum set or in the deployment). If this element is empty, the
+                maximum set will be assigned an empty set of permissions and will thus reject any permissions specified
+                in the minimum set or in the deployment. In this case, the empty set can be used to enforce a policy
+                where deployments cannot be assigned any permissions.
+
+                NOTE: the behavior regarding empty maximum sets has changed in version 3.0. In previous versions of the
+                subsystem an empty maximum set was handled the same way as an undefined set and that caused caused the
+                maximum set to resolve to a set containing java.security.AllPermission.
                         ]]>
             </xs:documentation>
         </xs:annotation>
@@ -89,6 +100,7 @@
                 * class: the permission fully-qualified class name (e.g. java.util.PropertyPermission).
                 * name: the name  of permission target (e.g. java.* for a PropertyPermission).
                 * actions: the permission actions (e.g. read, write)
+                * module: the name of the module that contains the permission class.
                         ]]>
             </xs:documentation>
         </xs:annotation>

--- a/security-manager/src/main/resources/subsystem-templates/security-manager.xml
+++ b/security-manager/src/main/resources/subsystem-templates/security-manager.xml
@@ -2,7 +2,7 @@
 <!--  See src/resources/configuration/ReadMe.txt for how the configuration assembly works -->
 <config>
    <extension-module>org.wildfly.extension.security.manager</extension-module>
-    <subsystem xmlns="urn:jboss:domain:security-manager:1.0">
+    <subsystem xmlns="urn:jboss:domain:security-manager:3.0">
         <deployment-permissions>
             <maximum-set>
                 <permission class="java.security.AllPermission"/>

--- a/security-manager/src/test/java/org/wildfly/extension/security/manager/SecurityManagerSubsystem10TestCase.java
+++ b/security-manager/src/test/java/org/wildfly/extension/security/manager/SecurityManagerSubsystem10TestCase.java
@@ -25,18 +25,18 @@ package org.wildfly.extension.security.manager;
 import java.io.IOException;
 
 import org.jboss.as.subsystem.test.AbstractSubsystemBaseTest;
+import org.jboss.as.subsystem.test.KernelServices;
 
 /**
- * Security Manager subsystem tests.
+ * Security Manager version 1.0 subsystem tests.
  *
  * @author <a href="sguilhen@jboss.com">Stefan Guilhen</a>
  */
-public class SecurityManagerSubsystemTestCase extends AbstractSubsystemBaseTest {
+public class SecurityManagerSubsystem10TestCase extends AbstractSubsystemBaseTest {
 
-    public SecurityManagerSubsystemTestCase() {
+    public SecurityManagerSubsystem10TestCase() {
         super(Constants.SUBSYSTEM_NAME, new SecurityManagerExtension());
     }
-
 
     @Override
     protected String getSubsystemXml() throws IOException {
@@ -49,9 +49,7 @@ public class SecurityManagerSubsystemTestCase extends AbstractSubsystemBaseTest 
     }
 
     @Override
-    protected String[] getSubsystemTemplatePaths() throws IOException {
-        return new String[] {
-                "/subsystem-templates/security-manager.xml"
-        };
+    protected KernelServices standardSubsystemTest(String configId, boolean compareXml) throws Exception {
+        return super.standardSubsystemTest(configId, false);
     }
 }

--- a/security-manager/src/test/java/org/wildfly/extension/security/manager/SecurityManagerSubsystem30TestCase.java
+++ b/security-manager/src/test/java/org/wildfly/extension/security/manager/SecurityManagerSubsystem30TestCase.java
@@ -1,0 +1,97 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2013, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.wildfly.extension.security.manager;
+
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.SUBSYSTEM;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+import java.io.IOException;
+
+import org.jboss.as.controller.ModelVersion;
+import org.jboss.as.controller.PathAddress;
+import org.jboss.as.controller.PathElement;
+import org.jboss.as.model.test.FailedOperationTransformationConfig;
+import org.jboss.as.model.test.ModelTestControllerVersion;
+import org.jboss.as.model.test.ModelTestUtils;
+import org.jboss.as.subsystem.test.AbstractSubsystemBaseTest;
+import org.jboss.as.subsystem.test.AdditionalInitialization;
+import org.jboss.as.subsystem.test.KernelServices;
+import org.jboss.as.subsystem.test.KernelServicesBuilder;
+import org.junit.Test;
+
+/**
+ * Security Manager version 3.0 subsystem tests.
+ *
+ * @author <a href="sguilhen@jboss.com">Stefan Guilhen</a>
+ */
+public class SecurityManagerSubsystem30TestCase extends AbstractSubsystemBaseTest {
+
+    public SecurityManagerSubsystem30TestCase() {
+        super(Constants.SUBSYSTEM_NAME, new SecurityManagerExtension());
+    }
+
+
+    @Override
+    protected String getSubsystemXml() throws IOException {
+        return readResource("security-manager-3.0.xml");
+    }
+
+    @Override
+    protected String getSubsystemXsdPath() throws Exception {
+        return "schema/wildfly-security-manager_3_0.xsd";
+    }
+
+    @Override
+    protected String[] getSubsystemTemplatePaths() throws IOException {
+        return new String[] {
+                "/subsystem-templates/security-manager.xml"
+        };
+    }
+
+    @Test
+    public void testTransformersEAP70() throws Exception {
+        testTransformers(ModelTestControllerVersion.EAP_7_0_0);
+    }
+
+    private void testTransformers(ModelTestControllerVersion controllerVersion) throws Exception {
+
+        KernelServicesBuilder builder = createKernelServicesBuilder(AdditionalInitialization.MANAGEMENT);
+        ModelVersion version = ModelVersion.create(2, 0, 0);
+        builder.createLegacyKernelServicesBuilder(AdditionalInitialization.MANAGEMENT, controllerVersion, version)
+                .addMavenResourceURL("org.jboss.eap:wildfly-security-manager:" + controllerVersion.getMavenGavVersion());
+
+        KernelServices mainServices = builder.build();
+        assertTrue(mainServices.isSuccessfulBoot());
+        KernelServices legacyServices = mainServices.getLegacyServices(version);
+        assertNotNull(legacyServices);
+        assertTrue(legacyServices.isSuccessfulBoot());
+
+        PathAddress subsystemAddress = PathAddress.pathAddress(PathElement.pathElement(SUBSYSTEM, getMainSubsystemName()));
+        ModelTestUtils.checkFailedTransformedBootOperations(mainServices, version,
+                builder.parseXmlResource("security-manager-3.0.xml"),
+                new FailedOperationTransformationConfig()
+                        .addFailedAttribute(PathAddress.pathAddress(subsystemAddress, PathElement.pathElement(Constants.DEPLOYMENT_PERMISSIONS)),
+                                FailedOperationTransformationConfig.REJECTED_RESOURCE));
+    }
+}

--- a/security-manager/src/test/resources/org/wildfly/extension/security/manager/security-manager-3.0.xml
+++ b/security-manager/src/test/resources/org/wildfly/extension/security/manager/security-manager-3.0.xml
@@ -1,0 +1,10 @@
+<subsystem xmlns="urn:jboss:domain:security-manager:3.0">
+    <deployment-permissions>
+        <minimum-set>
+            <permission class="java.io.FilePermission" name="/tmp" actions="read,write"/>
+            <permission class="java.io.FilePermission" name="/opt" actions="read"/>
+            <permission class="java.lang.RuntimePermission" name="createClassLoader"/>
+        </minimum-set>
+        <maximum-set/>
+    </deployment-permissions>
+</subsystem>

